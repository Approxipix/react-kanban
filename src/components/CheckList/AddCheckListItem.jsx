import React, { Component } from "react";
import { connect } from "react-redux";
import styled from 'styled-components'
import { bindActionCreators } from 'redux';
import { addCheckListItem } from '../../redux/rootReducer/actions';
import { Input, TextArea, SubmitButton, CancelButton } from '../BaseComponent';
import ClickOutside from "../ClickOutside";

const Wrapper = styled.div`
  position: relative;
  padding: .75rem .5rem;
  background-color: #ddd;
  border-radius: .4rem
`;

const Actions = styled.div`
  display: flex;
  align-items: center;
`;

const Form = styled.form``;

const AddButton = styled.button`
  text-align: left;
  width: 100%;
  margin-top: .5rem;
  padding: .75rem;
  font-size: 1rem; 
  color: #444;
  font-weight: 600;
  transition: background-color .1s;
  cursor: pointer;
  &:hover {
    background-color: #ccc;
  }
`;

class AddCheckListItem extends Component {
  constructor(props) {
    super(props);
    this.state = {
      isOpened: false,
      title: '',
    };
  }

  toggleOpened = () => {
    this.setState({
      isOpened: !this.state.isOpened,
      title: '',
    })
  };

  handleChange = (key, value) => {
    this.setState({
      ...this.state,
      [key]: value,
    });
  };

  handleSubmit = event => {
    event.preventDefault();
    const { title } = this.state;
    if (!title)  return;
    this.props.actions.addCheckListItem({
      checkListId: this.props.checkListId,
      checkListTitle: title,
    });
    this.toggleOpened()
  };

  render = () => {
    const { isOpened, title } = this.state;
    return isOpened ? (
      <ClickOutside toggleOpened={this.toggleOpened}>
        <Wrapper>
          <Form id="card-add-form">
            <Input
              autoFocus
              type="text"
              placeholder="Enter a card title..."
              value={title}
              onChange={(e) => this.handleChange('title', e.target.value)}
              spellCheck={false}
            />
            <Actions>
              <SubmitButton onClick={(e) => this.handleSubmit(e)} disabled={!title}>
                Save
              </SubmitButton>
              or
              <CancelButton onClick={() => this.toggleOpened()}>
                cancel
              </CancelButton>
            </Actions>
          </Form>
        </Wrapper>
      </ClickOutside>
    ) : (
      <AddButton onClick={() => this.toggleOpened()}>
        + Add a item
      </AddButton>
    );
  };
}

function mapDispatchToProps(dispatch) {
  return {
    actions: bindActionCreators({
      addCheckListItem: addCheckListItem,
    }, dispatch)
  };
}


export default connect(null, mapDispatchToProps)(AddCheckListItem);
