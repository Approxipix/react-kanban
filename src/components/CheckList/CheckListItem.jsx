import React, { Component } from 'react';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import { updateCheckListItem, deleteCheckListItem } from '../../redux/boardReducer/actions';
import styled from 'styled-components'
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { CheckBox } from '../BaseComponent';

const Item = styled.div`
  position: relative;
  display: flex;
  margin-bottom: 10px;
  line-height: 36px;
  border-bottom: 1px solid rgba(0, 0, 0, 0.2);
`;

const Desc = styled.label`
  display: flex;
  align-items: center;
  width: 100%;
  padding-right: 40px;
  cursor: pointer;
`;

const Text = styled.div`
  font-size: 16px;
  cursor: pointer;
  text-decoration: ${props => props.status ? 'line-through' : 'none'}
  color: ${props => props.status ? '#ccc' : 'initial'}
  font-style: ${props => props.status ? 'italic' : 'normal'}
`;

const Check = styled.input`
  display: none;
`;


const Button = styled.button`
  position: absolute;
  right: 0;
  top: 50%;
  transform: translateY(-50%);
  padding: 7px;
  color: #afafaf;
  line-height: 0;
  border: 0;
  background-color: transparent;
  outline: none;
  cursor: pointer;
  transition: all 0.2s ease-in;
  
  &:hover {
    color: #ccc;
  }
`;


class CheckListItem extends Component {

  handleDelete = () => {
    this.props.actions.deleteCheckListItem({
      cardIndex: this.props.cardIndex,
      checkBoxItemIndex: this.props.checkBoxItemIndex,
      listIndex: this.props.listIndex,
      checkBoxIndex: this.props.checkBoxIndex,
    })
  };

  handleStatus = () => {
    this.props.actions.updateCheckListItem({
      cardIndex: this.props.cardIndex,
      checkBoxItemIndex: this.props.checkBoxItemIndex,
      listIndex: this.props.listIndex,
      checkBoxIndex: this.props.checkBoxIndex,
      status: !this.props.status,
    })
  };

  render() {
    const { status, description } = this.props;
    if (this.props.hide && status) return null;
    return (
      <Item>
        <Desc onClick={(e) => e.stopPropagation() }>
          <CheckBox status={status}>
            <span>
              <svg width="12px" height="10px" viewBox="0 0 12 10">
                <polyline points="1.5 6 4.5 9 10.5 1" />
              </svg>
            </span>
          </CheckBox>
          <Check
            type="checkbox"
            onChange={() => this.handleStatus()}
            checked={status}
          />
          <Text status={status}>{description}</Text>
        </Desc>
        <Button type="button"
                onClick={() => this.handleDelete()}
        >
          delete
        </Button>
      </Item>
    );
  }
}

function mapStateToProps(state, ownProps) {
  const { rootReducer } = state;
  const board = rootReducer.boards[rootReducer.currentBoardIndex];
  const list = board.lists[ownProps.listIndex];
  const card = list.cards[ownProps.cardIndex];
  const items = card.checkLists[ownProps.checkBoxIndex].items[ownProps.checkBoxItemIndex];
  const status = card.checkLists[ownProps.checkBoxIndex].items[ownProps.checkBoxItemIndex].status;
  const description = card.checkLists[ownProps.checkBoxIndex].items[ownProps.checkBoxItemIndex].description;

  return {
    items: items,
    status: status,
    description: description
  }

}

function mapDispatchToProps(dispatch) {
  return {
    actions: bindActionCreators({
      updateCheckListItem: updateCheckListItem,
      deleteCheckListItem: deleteCheckListItem,
    }, dispatch)
  };
}

export default connect(mapStateToProps, mapDispatchToProps)(CheckListItem);
