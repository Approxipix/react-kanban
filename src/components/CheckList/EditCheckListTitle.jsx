import React, { Component } from "react";
import { connect } from "react-redux";
import styled from 'styled-components'
import { bindActionCreators } from 'redux';
import { editCheckListTitle } from '../../redux/rootReducer/actions';
import ClickOutside from "../ClickOutside";

const Form = styled.form`
  width: 90%;
`;

const Input = styled.input`
  padding: .6rem .5rem .5rem;
  width: 100%;
  border: none;
  font-weight: bold;
  box-shadow: none;
  font-size: 1rem;
  color: #40424b;
`;

class EditCheckListTitle extends Component {
  constructor(props) {
    super(props);
    this.state = {
      title: props.checkList.title || '',
    };
  }

  handleChange = (key, value) => {
    this.setState({
      ...this.state,
      [key]: value,
    });
  };

  handleSubmit = event => {
    event.preventDefault();
    const { title } = this.state;
    const { actions } = this.props;
    if (!title)  return;
    actions.editCheckListTitle({
      cardIndex: this.props.cardIndex,
      listIndex: this.props.listIndex,
      checkBoxIndex: this.props.checkBoxIndex,
      title: title,
    });
    this.props.toggleEditing()
  };

  render = () => {
    const { title } = this.state;
    return (
      <ClickOutside toggleOpened={this.props.toggleEditing}>
        <Form onSubmit={this.handleSubmit}>
          <Input
            autoFocus
            type="text"
            placeholder="Enter list title..."
            value={title}
            onChange={(e) => this.handleChange('title', e.target.value)}
            spellCheck={false}
          />
        </Form>
      </ClickOutside>
    );
  };
}

function mapStateToProps(state, ownProps) {
  const { rootReducer } = state;
  const board = rootReducer.boards[rootReducer.currentBoardIndex];
  const list = board.lists[ownProps.listIndex];
  const card = list.cards[ownProps.cardIndex];
  const checkList = card.checkLists[ownProps.checkBoxIndex];
  return {

    checkList: checkList
  }
}

function mapDispatchToProps(dispatch) {
  return {
    actions: bindActionCreators({
      editCheckListTitle: editCheckListTitle,
    }, dispatch)
  };
}


export default connect(mapStateToProps, mapDispatchToProps)(EditCheckListTitle);
