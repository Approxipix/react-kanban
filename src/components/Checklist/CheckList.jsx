import React, { Component } from 'react';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import { deleteCheckList } from '../../redux/checkListReducer/actions';
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import CheckListProgress from './CheckListProgress'
import CheckListItemAdd from './CheckListItem/CheckListItemAdd'
import CheckListTitle from './CheckListTitle/CheckListTitle'
import CheckListItem from './CheckListItem/CheckListItem'
import styled from 'styled-components'

const Container = styled.div`
  position: relative;
`;

const Header = styled.div`
  display: grid;
  grid-template-columns: 1fr auto;
  grid-gap: .5rem;
  align-items: flex-start;
  position: relative;
  margin-bottom: .5rem;
  padding-left: calc(2rem - .5rem);
`;

const Icon = styled.div`
  position: absolute;
  left: 0;
  top: .45rem;
  font-size: 1rem;
`;

const CheckListContainer = styled.div`
  margin-bottom: .5rem;
`;

const Actions = styled.div`
  display: flex;
`;

const Button = styled.button`
  padding: .5rem 1rem;
  color: #172b4d;
  background-color: rgba(9, 30, 66, .08);
  border-radius: 3px;
  transition: background-color .2s ease-in;
  &:hover {
    background-color: rgba(9,30,66,.13);
  }
  &:not(:last-child) {
    margin-right: .5rem;
  }
`;

class CheckList extends Component {
  constructor(props) {
    super(props);
    this.state = {
      hideCompletedItems: false,
    }
  }

  deleteCheckList = () => {
    this.props.actions.deleteCheckList({
      cardId: this.props.cardId,
      checkListId: this.props.checkList._checkListId,
    })
  };

  toggleHideCompletedItems = () => {
    this.setState({
      hideCompletedItems: !this.state.hideCompletedItems
    })
  };

  render() {
    const { checkList } = this.props;
    const { hideCompletedItems } = this.state;
    const doneItems = checkList.items.filter(items => !!items.status).length;
    return (
      <Container>
        <Header>
          <Icon>
            <FontAwesomeIcon icon="check-square" />
          </Icon>
          <CheckListTitle
            checkListTitle={checkList.title}
            checkListId={checkList._checkListId}
          />
          <Actions>
            {!!doneItems && (
              <Button onClick={() => this.toggleHideCompletedItems()}>
                {!hideCompletedItems
                  ? 'Hide completed items'
                  : `Show checked items (${doneItems})`
                }
              </Button>
            )}
            <Button onClick={() => this.deleteCheckList()}>
              Delete
            </Button>
          </Actions>
        </Header>
        <CheckListProgress items={checkList.items} />
        <CheckListContainer>
          {checkList.items.map((item, index) => (
            <CheckListItem
              key={index}
              index={index}
              status={item.status}
              hideCompletedItems={hideCompletedItems}
              description={item.description}
              checkListId={checkList._checkListId}
            />
          ))}
        </CheckListContainer>
        <CheckListItemAdd checkListId={checkList._checkListId}/>
      </Container>
    );
  }
}

function mapStateToProps(state, ownProps) {
  return {
    checkList: state.checkListReducer[ownProps.chekListId],
  }
}

function mapDispatchToProps(dispatch) {
  return {
    actions: bindActionCreators({
      deleteCheckList: deleteCheckList,
    }, dispatch)
  };
}

export default connect(mapStateToProps, mapDispatchToProps)(CheckList);
