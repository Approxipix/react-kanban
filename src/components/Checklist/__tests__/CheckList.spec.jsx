import React from 'react';
import configureStore from 'redux-mock-store';
import { shallow } from 'enzyme';
import toJson from "enzyme-to-json";
import CheckList from '../CheckList';

const mockStore = configureStore([]);

describe('<CheckList>', () => {
  let store;
  let props;
  let state;
  let component;

  beforeEach(() => {
    store = mockStore({
      checkListReducer: {
        CheckListID: {
          _checkListId: 'CheckListID',
          title: 'CheckList Title',
          items: [
            { description: 'CheckList Item1 Title', status: false },
            { description: 'CheckList Item2 Title', status: true },
          ]
        }
      }
    });
    store.dispatch = jest.fn();
    props = {
      cardId: 'CardID',
      checkListId: 'CheckListID',
    };
    state = {
      hideCompletedItems: false,
    };

    const wrapper = shallow(<CheckList store={store} {...props} />);
    component = wrapper.find('CheckList').dive();
    component.setState({...state});
  });

  it('should render with given state from Redux store and correct props', () => {
    expect(toJson(component)).toMatchSnapshot();
  });

  it('should render a CheckListItem component for each items in checklist', () => {
    expect(component.find('CheckListContainer').children().length).toEqual(2)
  });

  it('should call toggleHideCompletedItems after click button', () => {
    const spy = jest.spyOn(component.instance(), "toggleHideCompletedItems");

    component.find('Button').last().simulate('click');

    expect(spy).toHaveBeenCalledTimes(1);
  });

  it('should call setState on hideCompletedItems', () => {
    const expectedState = {
      hideCompletedItems: true,
    };

    component.instance().toggleHideCompletedItems();

    expect(component.state()).toEqual(expectedState)
  });

  it('should render correct button text depending on the state', () => {
    expect(component.find('Button').last().text()).toEqual('Hide completed items');

    component.instance().toggleHideCompletedItems();

    expect(component.find('Button').last().text()).toEqual('Show checked items (1)');
  });

  it('should dispatch an action on delete button click', () => {
    const expectedAction = {
      type: 'DELETE_CHECKLIST',
      payload: {
        cardId: 'CardID',
        checkListId: 'CheckListID',
      }
    };

    component.find('Button').first().simulate('click');

    expect(store.dispatch).toHaveBeenCalledTimes(1);
    expect(store.dispatch).toHaveBeenCalledWith(expectedAction);
  });
});
