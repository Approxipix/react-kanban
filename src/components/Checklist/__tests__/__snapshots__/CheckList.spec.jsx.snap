// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CheckList> should render with given state from Redux store and correct props 1`] = `
<styled.div>
  <styled.div>
    <styled.div>
      <FontAwesomeIcon
        border={false}
        className=""
        fixedWidth={false}
        flip={null}
        icon="check-square"
        inverse={false}
        listItem={false}
        mask={null}
        pull={null}
        pulse={false}
        rotation={null}
        size={null}
        spin={false}
        swapOpacity={false}
        symbol={false}
        title=""
        transform={null}
      />
    </styled.div>
    <CheckListTitle
      checkListId="CheckListID"
      checkListTitle="CheckList Title"
    />
    <Button
      onClick={[Function]}
    >
      Delete
    </Button>
  </styled.div>
  <CheckListProgress
    items={
      Array [
        Object {
          "description": "CheckList Item1 Title",
          "status": false,
        },
        Object {
          "description": "CheckList Item2 Title",
          "status": true,
        },
      ]
    }
  />
  <CheckListContainer>
    <Connect(CheckListItem)
      checkListId="CheckListID"
      description="CheckList Item1 Title"
      hideCompletedItems={false}
      index={0}
      key="0"
      status={false}
    />
    <Connect(CheckListItem)
      checkListId="CheckListID"
      description="CheckList Item2 Title"
      hideCompletedItems={false}
      index={1}
      key="1"
      status={true}
    />
  </CheckListContainer>
  <styled.div>
    <Connect(CheckListItemAdd)
      checkListId="CheckListID"
    />
    <Button
      onClick={[Function]}
    >
      Hide completed items
    </Button>
  </styled.div>
</styled.div>
`;
